File: src/adventofcode2018/Day20.java
package adventofcode2018;

import java.awt.Point;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;

import utils.Input;

/**
 * https://adventofcode.com/2018/day/20
 */
public class Day20 {

    static int route(Point pt, String path, int steps, Map<Point, Integer> map) {
        int mySteps = 0;
        Point myPoint = pt;
        for (int i=0; ; i++) {
            Point next = null;
            char c = path.charAt(i);
            if (c == 'E') {
                next = new Point(myPoint.x + 1, myPoint.y);
            } else if (c == 'W') {
                next = new Point(myPoint.x - 1, myPoint.y);
            } else if (c == 'N') {
                next = new Point(myPoint.x, myPoint.y - 1);
            } else if (c == 'S') {
                next = new Point(myPoint.x, myPoint.y + 1);
            } else if (c == '(') {
                i += route(myPoint, path.substring(i+1), steps + mySteps, map);
                continue;
            } else if (c == ')' || c =='$') {
                return i + 1;
            } else if (c == '|') {
                myPoint = pt;
                mySteps = 0;
            }
            if (next != null) {
                myPoint = next;
                mySteps++;
                if (!map.containsKey(next)) {
                    map.put(next, steps + mySteps);
                }
            }
        }
    }

    public static void main(String[] args) throws Exception {
        String input = new Input(2018, 20).strings().get(0);
        Map<Point, Integer> map = new HashMap<>();
        route(new Point(0, 0), input, 0, map);

        System.err.println(map.values().stream().max(Comparator.naturalOrder()).get());
        System.err.println(map.values().stream().filter(x -> x >= 1000).count());
        System.err.println();
    }
}

File: src/Python/2018/Day20.py

from utils import Input, printResult

# https://adventofcode.com/2018/day/20

def route(pt, path, steps, map):
    mySteps = 0
    myPoint = pt
    i = -1
    deltas = {"E": [1, 0], "W": [-1, 0], "N": [0, -1], "S": [0, 1]}
    while True:
        i += 1
        c = path[i]
        if c in deltas:
            myPoint = (myPoint[0] + deltas[c][0], myPoint[1] + deltas[c][1])
            mySteps += 1
            if myPoint not in map:
                map[myPoint] = steps + mySteps
        elif c == "(":
            i += route(myPoint, path[i+1:], steps + mySteps, map)
            continue
        elif c in [")", "$"]:
            return i + 1
        elif c == "|":
            myPoint = pt
            mySteps = 0

input = Input(2018, 20).lines()[0]
map = {}
route((0, 0), input, 0, map)

printResult(1, max(map.values()))
printResult(2, sum(1 for _ in filter(lambda x: x >= 1000, map.values())))
